src/fsharp/Optimizer.fs

```diff
@@ -1575,7 +1575,12 @@ let MakeStructuralBindingTemp (v: Val) i (arg: Expr) argTy =
     let name = v.LogicalName + "_" + string i
     let v, ve = mkCompGenLocal arg.Range name argTy
     ve, mkCompGenBind v arg
-           
+
+let MakeMutableStructuralBindingForTupleElement (v: Val) i (arg: Expr) argTy =
+    let name = sprintf "%s_%d_%s" v.LogicalName i PrettyNaming.tempTupleElementAssignmentTargetName
```

Comment:
If I understand correctly, these temporaries should always be removed within Optimizer.fs, even in debug code?

Reply:
I don't think so. Given this:

```fsharp
let x () =
    let a, b =
        sideEffect ()
        if cond () then
            let v = "yep"
            let v2 = if cond () then 1 else 3
            v, v2
        else
            "", f ()
    a, b
```

no tuples are created on the right hand side, but those temporaries will still be used to create a tuple in the let body. Yes, the optimization is a bit pointless here, but I reckon it would be more expensive to detect whether the temporaries will be eliminated in the end, rather than just rewrite the right hand side.

Reply:
Perhaps not pointless at all. [This is what happens today.](https://sharplab.io/#v2:DYLgZgzgPg2gPAZQJ4QC4FMC2A6ASgVwDtUBLTdbAYQHtMAHE4dAJwRYDcSBjdCbAWXSoAFtQAmASXrAAFMjRY8RUuSq0GTVh268BQ0ZOkB5OqWqE+AOWoTCwEoQcBzAJQA+ALoBYAFBNUAARc5mIBMi4BALwBqMz46L6+8PIYOATEZBQ09IwsbMycPHyCIuJSdLIpiukqWeq5WgU6xfplxqYk5lY2dg7O7t5+QgFgYRHRAMyJQ4EAHmNRvgHLAf4BAIYANAEARos+K4cBJKPBhKHhMcLohEtH92vsUQEAREjodC9394ePAEzPE5BEILEQ3AIARgC6GAEHQASmBx+h3Y23Yf2+RxhcMxyJeL22o3CmK2uyAA)